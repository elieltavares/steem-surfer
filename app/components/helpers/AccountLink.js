import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { getAccount } from '../../backend/steem-client';

export const makePath = username => `/@${username}`;

class AccountLink extends Component {
  goProfile = async () => {
    const { username, history, actions, onClick, afterClick } = this.props;

    onClick();

    let { accountData } = this.props;
    const { setVisitingAccount } = actions;

    if (!accountData) {
      try {
        accountData = await getAccount(username);
      } catch (err) {
        accountData = null;
      }
    }

    if (accountData) {
      setVisitingAccount(accountData);
    }

    history.push(makePath(username));

    afterClick();
  };

  render() {
    const { children } = this.props;

    return React.cloneElement(children, {
      onClick: this.goProfile
    });
  }
}

AccountLink.defaultProps = {
  accountData: null,
  onClick: () => {},
  afterClick: () => {}
};

AccountLink.propTypes = {
  children: PropTypes.element.isRequired,
  username: PropTypes.string.isRequired,
  history: PropTypes.instanceOf(Object).isRequired,
  accountData: PropTypes.instanceOf(Object),
  onClick: PropTypes.func,
  afterClick: PropTypes.func,
  actions: PropTypes.shape({
    setVisitingAccount: PropTypes.func.isRequired
  }).isRequired
};

export default AccountLink;
