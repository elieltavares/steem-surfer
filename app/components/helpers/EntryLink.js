import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { getContent } from '../../backend/steem-client';

export const makePath = (category, author, permlink, toReplies = false) =>
  `/${category}/@${author}/${permlink}${toReplies ? '#replies' : ''}`;

class EntryLink extends Component {
  goEntry = async () => {
    let { entry } = this.props;
    const { history, actions, toReplies, markAsRead } = this.props;

    if (markAsRead) {
      markAsRead();
    }

    if (!entry) {
      const { author, permlink } = this.props;
      entry = await getContent(author, permlink);
    }

    if (entry) {
      const { category, author, permlink } = entry;
      actions.setVisitingEntry(entry);
      history.push(makePath(category, author, permlink, toReplies));
    }
  };

  render() {
    const { children } = this.props;

    return React.cloneElement(children, {
      onClick: this.goEntry
    });
  }
}

EntryLink.defaultProps = {
  entry: null,
  toReplies: false,
  markAsRead: null
};

EntryLink.propTypes = {
  children: PropTypes.element.isRequired,
  history: PropTypes.instanceOf(Object).isRequired,
  entry: PropTypes.shape({
    category: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    permlink: PropTypes.string.isRequired
  }),
  author: PropTypes.string.isRequired,
  permlink: PropTypes.string.isRequired,
  toReplies: PropTypes.bool,
  actions: PropTypes.shape({
    setVisitingEntry: PropTypes.func.isRequired
  }).isRequired,
  markAsRead: PropTypes.func
};

export default EntryLink;
