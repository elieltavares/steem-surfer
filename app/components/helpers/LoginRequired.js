import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';

import { Modal } from 'antd';
import { FormattedMessage, FormattedHTMLMessage } from 'react-intl';

import Login from '../dialogs/Login';

export const checkLogin = (activeAccount, requiredKeys) => {
  // Steem connect login.
  if (activeAccount && activeAccount.type === 'sc') {
    return true;
  }

  // Traditional login. Check all required keys exists in user keys.
  if (
    activeAccount &&
    activeAccount.type === 's' &&
    requiredKeys.every(e => activeAccount.keys[e])
  ) {
    return true;
  }

  return false;
};

class LoginRequired extends Component {
  constructor(props) {
    super(props);

    this.state = {
      loginModalVisible: false
    };
  }

  showLoginModal = () => {
    this.setState({
      loginModalVisible: true
    });

    setTimeout(() => {
      const { onDialogOpen } = this.props;
      onDialogOpen();
    }, 200);
  };

  onLoginModalCancel = () => {
    this.setState({
      loginModalVisible: false
    });
  };

  onLoginSuccess = () => {
    this.setState({
      loginModalVisible: false
    });

    setTimeout(() => {
      if (!this.check()) {
        this.showLoginModal();
      }
    }, 400);
  };

  afterModalClose = () => {
    const { onDialogClose } = this.props;
    onDialogClose();
  };

  check = () => {
    const { activeAccount, requiredKeys } = this.props;

    return checkLogin(activeAccount, requiredKeys);
  };

  render() {
    const { children } = this.props;

    if (this.check()) {
      return children;
    }

    const { activeAccount, requiredKeys } = this.props;
    const { loginModalVisible } = this.state;

    // Default message
    let loginMsg = <FormattedMessage id="login-required.default" />;

    if (activeAccount) {
      // More specific, key based message
      loginMsg =
        requiredKeys.length > 1 ? (
          <FormattedHTMLMessage
            id="login-required.keys-required"
            values={{ keys: requiredKeys }}
          />
        ) : (
          <FormattedHTMLMessage
            id="login-required.key-required"
            values={{ key: requiredKeys }}
          />
        );
    }

    // Clone child element rewriting onclick event
    const newChild = React.cloneElement(children, {
      onClick: () => {
        this.showLoginModal();
      }
    });

    return (
      <Fragment>
        {newChild}

        <Modal
          visible={loginModalVisible}
          onCancel={this.onLoginModalCancel}
          footer={false}
          width="500px"
          closable={false}
          destroyOnClose
          centered
          afterClose={this.afterModalClose}
        >
          <Login
            {...this.props}
            onSuccess={this.onLoginSuccess}
            loginMsg={loginMsg}
          />
        </Modal>
      </Fragment>
    );
  }
}

LoginRequired.defaultProps = {
  activeAccount: null,
  onDialogOpen: () => {},
  onDialogClose: () => {},
  requiredKeys: []
};

LoginRequired.propTypes = {
  children: PropTypes.element.isRequired,
  activeAccount: PropTypes.instanceOf(Object),
  requiredKeys: PropTypes.arrayOf(PropTypes.string),
  onDialogOpen: PropTypes.func,
  onDialogClose: PropTypes.func
};

export default LoginRequired;
