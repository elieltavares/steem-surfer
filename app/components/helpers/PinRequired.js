import React, { Component, Fragment } from 'react';
import { Modal } from 'antd';

import PropTypes from 'prop-types';
import PinConfirm from '../dialogs/PinConfirm';

class PinRequired extends Component {
  constructor(props) {
    super(props);

    this.state = {
      show: false
    };
  }

  onPinInvalidated = () => {
    this.setState({ show: false });
  };

  onConfirmPinSuccess = value => {
    const { onSuccess } = this.props;
    this.setState({ show: false }, () => {
      onSuccess(value);
    });
  };

  cancelFn = () => {
    this.setState({ show: false });
  };

  render() {
    const { children } = this.props;
    const { show } = this.state;

    // Clone child element rewriting onclick event
    const newChild = React.cloneElement(children, {
      onClick: () => {
        this.setState({ show: true });
      }
    });

    return (
      <Fragment>
        {show && (
          <Modal
            footer={null}
            closable
            onCancel={this.cancelFn}
            keyboard
            visible
            width="500px"
            centered
            destroyOnClose
          >
            <PinConfirm
              {...this.props}
              onSuccess={this.onConfirmPinSuccess}
              onInvalidate={this.onPinInvalidated}
            />
          </Modal>
        )}
        {newChild}
      </Fragment>
    );
  }
}

PinRequired.defaultProps = {
  onSuccess: () => {}
};

PinRequired.propTypes = {
  children: PropTypes.element.isRequired,
  history: PropTypes.instanceOf(Object).isRequired,
  onSuccess: PropTypes.func
};

export default PinRequired;
