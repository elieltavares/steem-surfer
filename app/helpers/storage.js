const PREFIX = 'surfer2';

export const getByPrefix = prefix => {
  const prefKey = `${PREFIX}_${prefix}`;

  return Object.keys(localStorage)
    .filter(key => key.indexOf(prefKey) === 0)
    .map(key => JSON.parse(localStorage[key]));
};

export const getItem = (k, def = null) => {
  const key = `${PREFIX}_${k}`;
  return localStorage.getItem(key)
    ? JSON.parse(localStorage.getItem(key))
    : def;
};

export const setItem = (k, v) => {
  const key = `${PREFIX}_${k}`;
  localStorage.setItem(key, JSON.stringify(v));
};

export const removeItem = k => {
  const key = `${PREFIX}_${k}`;
  localStorage.removeItem(key);
};

export const getVotingPercentage = username =>
  getItem(`voting_percentage_${username}`, 100);

export const setVotingPercentage = (username, val) => {
  setItem(`voting_percentage_${username}`, val);
};

export const getVotingPercentageNeg = username =>
  getItem(`voting_percentage_neg_${username}`, 100);

export const setVotingPercentageNeg = (username, val) => {
  setItem(`voting_percentage_neg_${username}`, val);
};

export const isEntryRead = (author, permlink) =>
  getItem(`read_flag_${author}_${permlink}`, false);

export const setEntryRead = (author, permlink) => {
  setItem(`read_flag_${author}_${permlink}`, true);
};

export const isReleasePostRead = ver => getItem(`release-post-${ver}`, false);

export const setReleasePostRead = ver => {
  setItem(`release-post-${ver}`, 1);
};

export const getRecentTransfers = () => getItem('recent-transfers', []);

export const appendToRecentTransfers = to => {
  const recent = getItem('recent-transfers', []);
  if (!recent.includes(to)) {
    const newRecent = [...new Set([to, ...recent])].slice(0, 100);
    setItem('recent-transfers', newRecent);
  }
};
