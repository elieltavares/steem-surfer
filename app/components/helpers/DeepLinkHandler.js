import React from 'react';
import PropTypes from 'prop-types';

import filters from '../../constants/filters.json';

import { makePath as makePathTag } from './TagLink';
import { makePath as makePathAccount } from './AccountLink';
import { makePath as makePathEntry } from './EntryLink';

export const deepLink2Obj = url => {
  let urlPart = url.split('://')[1];

  // remove last char if /
  if (urlPart.endsWith('/')) {
    urlPart = urlPart.substring(0, urlPart.length - 1);
  }

  const parts = urlPart.split('/');

  // witnesses
  if (parts.length === 1 && parts[0] === 'witnesses') {
    return { type: 'witnesses' };
  }

  // sps
  if (parts.length === 1 && parts[0] === 'sps') {
    return { type: 'sps' };
  }

  // filter
  if (parts.length === 1 && filters.includes(parts[0])) {
    return { type: 'filter', filter: parts[0] };
  }

  // filter with tag
  if (parts.length === 2 && filters.includes(parts[0])) {
    return { type: 'filter-tag', filter: parts[0], tag: parts[1] };
  }

  // account
  if (parts.length === 1 && parts[0].startsWith('@')) {
    return { type: 'account', account: parts[0].replace('@', '') };
  }

  // post without tag
  if (parts.length === 2 && parts[0].startsWith('@')) {
    return {
      type: 'post',
      cat: 'esteem',
      author: parts[0].replace('@', ''),
      permlink: parts[1]
    };
  }

  // post with tag
  if (parts.length === 3 && parts[1].startsWith('@')) {
    return {
      type: 'post',
      cat: parts[0],
      author: parts[1].replace('@', ''),
      permlink: parts[2]
    };
  }
};

class DeepLinkHandler extends React.Component {
  componentDidMount() {
    window.addEventListener('deep-link', this.handler);
  }

  componentWillUnmount() {
    window.removeEventListener('deep-link', this.handler);
  }

  handler = e => {
    const { history } = this.props;
    const { url } = e.detail;
    const obj = deepLink2Obj(url);

    let path;

    switch (obj.type) {
      case 'filter':
        path = `/${obj.filter}`;
        break;
      case 'filter-tag':
        path = makePathTag(obj.filter, obj.tag);
        break;
      case 'account':
        path = makePathAccount(obj.account);
        break;
      case 'post':
        path = makePathEntry(obj.cat, obj.author, obj.permlink);
        break;
      case 'sps':
        path = `/sps`;
        break;
      case 'witnesses':
        path = `/witnesses`;
        break;
      default:
        path = null;
        break;
    }

    if (path) {
      history.push(path);
    }
  };

  render() {
    return null;
  }
}

DeepLinkHandler.propTypes = {
  history: PropTypes.shape({}).isRequired
};

export default DeepLinkHandler;
