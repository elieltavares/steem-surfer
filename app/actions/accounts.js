import { encryptKey } from '../utils/crypto';
import { setItem, removeItem, getByPrefix } from '../helpers/storage';

export const ACCOUNT_ADDED = 'accounts/ACCOUNT_ADDED';
export const ACCOUNT_ADDED_SC = 'accounts/ACCOUNT_SC_ADDED';
export const ACCOUNT_UPDATED_SC = 'accounts/ACCOUNT_UPDATED_SC';
export const ACCOUNT_DELETED = 'accounts/ACCOUNT_DELETED';
export const ACCOUNTS_DELETED = 'accounts/ACCOUNTS_DELETED';

export const addAccount = (username, keys) => (dispatch, getState) => {
  const { global } = getState();

  const { pin } = global;

  // key encryption
  const eKeys = Object.assign(
    {},
    ...Object.keys(keys).map(k => ({
      [k]: keys[k] ? encryptKey(keys[k], pin) : null
    }))
  );

  const accountData = {
    type: 's',
    username,
    keys: eKeys
  };

  setItem(`account_${username}`, accountData);
  // setItem(`active_account`, username);
  dispatch(accountAdded(username, accountData));
};

export const addAccountSc = (
  username,
  accessToken,
  refreshToken,
  expiresIn,
  updated = false
) => (dispatch, getState) => {
  const { global } = getState();

  const { pin } = global;

  const accountData = {
    type: 'sc',
    username,
    accessToken: encryptKey(accessToken, pin),
    refreshToken: encryptKey(refreshToken, pin),
    expiresIn
  };

  setItem(`account_${username}`, accountData);
  if (!updated) {
    dispatch(accountAddedSc(username, accountData));
  } else {
    dispatch(accountUpdatedSc(username, accountData));
  }
};

export const deleteAccount = username => dispatch => {
  removeItem(`account_${username}`);

  dispatch(accountDeleted(username));
};

export const deleteAccounts = () => dispatch => {
  getByPrefix('account_').forEach(ac => {
    removeItem(`account_${ac.username}`);
  });

  dispatch(accountsDeleted());
};

/* action creators */

export const accountAdded = username => ({
  type: ACCOUNT_ADDED,
  payload: { username }
});

export const accountAddedSc = username => ({
  type: ACCOUNT_ADDED_SC,
  payload: { username }
});

export const accountUpdatedSc = (username, accountData) => ({
  type: ACCOUNT_UPDATED_SC,
  payload: { username, accountData }
});

export const accountDeleted = username => ({
  type: ACCOUNT_DELETED,
  payload: { username }
});

export const accountsDeleted = () => ({
  type: ACCOUNTS_DELETED,
  payload: {}
});
