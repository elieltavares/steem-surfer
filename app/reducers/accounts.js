import { getByPrefix } from '../helpers/storage';

import {
  ACCOUNT_ADDED,
  ACCOUNT_ADDED_SC,
  ACCOUNT_UPDATED_SC,
  ACCOUNT_DELETED,
  ACCOUNTS_DELETED
} from '../actions/accounts';

const defaultState = getByPrefix('account_');

export default function accounts(state = defaultState, action) {
  switch (action.type) {
    case ACCOUNT_ADDED:
    case ACCOUNT_ADDED_SC:
    case ACCOUNT_UPDATED_SC:
    case ACCOUNT_DELETED:
    case ACCOUNTS_DELETED: {
      return getByPrefix('account_');
    }
    default:
      return state;
  }
}
