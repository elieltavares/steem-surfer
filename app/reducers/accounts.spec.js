/* eslint-disable */
import accounts from './accounts';
import {setItem, removeItem} from '../helpers/storage';

import {
  ACCOUNT_ADDED,
  ACCOUNT_DELETED,
  ACCOUNT_ACTIVATED,
  ACTIVE_ACCOUNT_DATA_UPDATED
} from '../actions/accounts';

import deepFreeze from 'deep-freeze';

describe('accounts reducer', () => {
  let state = undefined;

  it('(1) default state', () => {
    state = accounts(state, {});

    deepFreeze(state);

    expect(state).toMatchSnapshot();
  });

  it('(2) account "foo" added ', () => {
    const username = 'foo';

    setItem(`account_${username}`, {username});

    const action = {
      type: ACCOUNT_ADDED,
      payload: {username}
    };

    deepFreeze(state);

    state = accounts(state, action);

    expect(state).toMatchSnapshot();
  });

  it('(3) account "bar" added ', () => {
    const username = 'bar';

    setItem(`account_${username}`, {username});

    const action = {
      type: ACCOUNT_ADDED,
      payload: {username}
    };

    deepFreeze(state);

    state = accounts(state, action);

    expect(state).toMatchSnapshot();
  });

  it('(4) account "baz" added', () => {
    const username = 'baz';

    setItem(`account_${username}`, {username});

    const action = {
      type: ACCOUNT_ADDED,
      payload: {username}
    };

    deepFreeze(state);

    state = accounts(state, action);

    expect(state).toMatchSnapshot();
  });

  it('(5) account "baz" deleted', () => {
    const username = 'baz';

    removeItem(`account_${username}`);

    const action = {
      type: ACCOUNT_DELETED,
      payload: {username}
    };

    deepFreeze(state);

    state = accounts(state, action);

    expect(state).toMatchSnapshot();
  });

  it('(7) account "foo" deleted', () => {
    const username = 'foo';

    removeItem(`account_${username}`);

    const action = {
      type: ACCOUNT_DELETED,
      payload: {username}
    };

    deepFreeze(state);

    state = accounts(state, action);

    expect(state).toMatchSnapshot();
  });

  it('(10) account "bar" added ', () => {
    const username = 'bar';

    setItem(`account_${username}`, {username});

    const action = {
      type: ACCOUNT_ADDED,
      payload: {username}
    };

    deepFreeze(state);

    state = accounts(state, action);

    expect(state).toMatchSnapshot();
  });
});
