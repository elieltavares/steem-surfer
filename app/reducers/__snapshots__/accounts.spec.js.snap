// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accounts reducer (1) default state 1`] = `Array []`;

exports[`accounts reducer (2) account "foo" added  1`] = `
Array [
  Object {
    "username": "foo",
  },
]
`;

exports[`accounts reducer (3) account "bar" added  1`] = `
Array [
  Object {
    "username": "foo",
  },
  Object {
    "username": "bar",
  },
]
`;

exports[`accounts reducer (4) account "baz" added 1`] = `
Array [
  Object {
    "username": "foo",
  },
  Object {
    "username": "bar",
  },
  Object {
    "username": "baz",
  },
]
`;

exports[`accounts reducer (5) account "baz" deleted 1`] = `
Array [
  Object {
    "username": "foo",
  },
  Object {
    "username": "bar",
  },
]
`;

exports[`accounts reducer (7) account "foo" deleted 1`] = `
Array [
  Object {
    "username": "bar",
  },
]
`;

exports[`accounts reducer (10) account "bar" added  1`] = `
Array [
  Object {
    "username": "bar",
  },
]
`;
